flocker/node/_loop.py

```diff
@@ -0,0 +1,344 @@
+    """
+    A rich input indicating the cluster status has been received from the
+    control service.
+
+    :ivar Deployment configuration: Desired cluster configuration.
+    :ivar Deployment state: Actual cluster state.
+    """
+
+
+class ClusterStatusStates(Names):
+    """
+    States of the cluster status state machine.
+    """
+    # The client is currently disconnected:
+    DISCONNECTED = NamedConstant()
+    # The client is connected, we don't know cluster status:
+    IGNORANT = NamedConstant()
+    # The client is connected and we know the cluster status:
+    KNOWLEDGEABLE = NamedConstant()
+    # The system is shut down:
+    SHUTDOWN = NamedConstant()
+
+
+class ClusterStatusOutputs(Names):
+    """
+    Outputs of the cluster status state machine.
+    """
+    # Store the AMP protocol instance connected to the server:
+    STORE_CLIENT = NamedConstant()
+    # Notify the convergence loop state machine of new cluster status:
+    UPDATE_STATUS = NamedConstant()
+    # Stop the convergence loop state machine:
+    STOP = NamedConstant()
+    # Disconnect the AMP client:
+    DISCONNECT = NamedConstant()
+
+
+class ClusterStatus(object):
+    """
+    World object for cluster state machine, executing the actions
+    indicated by the outputs.
+
+    :ivar AMP client: The latest AMP protocol instance to connect to the
+        control service. Initially ``None``.
+    """
+
+    def __init__(self, convergence_loop_fsm):
+        """
+        :param convergence_loop_fsm: An convergence loop FSM as output by
+            ``build_convergence_loop_fsm``.
+        """
+        self.convergence_loop_fsm = convergence_loop_fsm
+        self.client = None
+
+    def output_STORE_CLIENT(self, context):
+        self.client = context.client
+
+    def output_UPDATE_STATUS(self, context):
+        self.convergence_loop_fsm.receive(
+            _ClientStatusUpdate(client=self.client,
+                                configuration=context.configuration,
+                                state=context.state))
+
+    def output_STOP(self, context):
+        self.convergence_loop_fsm.receive(ConvergenceLoopInputs.STOP)
+
+    def output_DISCONNECT(self, context):
+        self.client.transport.loseConnection()
+
+
+def build_cluster_status_fsm(convergence_loop_fsm):
+    """
+    Create a new cluster status FSM.
+
+    :param convergence_loop_fsm: A convergence loop FSM as output by
+    ``build_convergence_loop_fsm``.
+    """
+    S = ClusterStatusStates
+    I = ClusterStatusInputs
+    O = ClusterStatusOutputs
+    table = TransitionTable()
+    # We may be shut down in any state, in which case we disconnect if
+    # necessary.
+    table = table.addTransitions(
+        S.DISCONNECTED, {
+            # Store the client, then wait for cluster status to be sent
+            # over AMP:
+            I.CLIENT_CONNECTED: ([O.STORE_CLIENT], S.IGNORANT),
+            I.SHUTDOWN: ([], S.SHUTDOWN),
```

Comment:
:question: While first looking at this, I wondered why this doesn't stop trying to connect the client. It looks like that happens outside of the state machine. Perhaps a comment *somewhere* (I don't know where) would be appropriate.